### ISSUE TITLE
Parlatype fails to launch, with a gstreamer and parlasphinx error

### ISSUE THREAD

**Issue by donnekgit** [gkarsay/parlatype#78]
### Installation and version

How did you install Parlatype?
From the Ubuntu 20.04 stock repositories.  The version is 1.6.2-1.  Other packages installed are gir1.2-parlatype-2.0, libparlatype2, parlatype-libreoffice-helpers.

### Your desktop environment
Ubuntu 20.04, running KDE Plasma as the desktop.

### Issue
Launching Parlatype fails, with a message that Parlatype needs GStreamer 1.x to run (Good Plugins) and it has been unable to load parlasphinx. But Gstreamer is installed - gstreamer1.0-plugins-good is v1.16.2,, and libsphinxbase3 is also installed.  The package worked fine in Ubuntu 18.04.  Starting from the terminal gives the same message.

### Other information

This seems to be related to issue 74:
https://github.com/gkarsay/parlatype/issues/74
though in that case the user was building a more recent version from a tarball.  It is likely, however, that most users would just use the Ubuntu rep version.  Similar issues were raised in 69 and 70 (both closed), but these were from a flathub install.

**Comment by gkarsay** [comment]
Thanks for your detailed report! I can't reproduce it installing Kubuntu 20.04 in a VM and installing Parlatype from stock repositories.

The missing plugin is provided by this file:
`/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgstparlasphinx.so` which is part of the libparlatype2 package. Do you have it on your system?

**Comment by evazzoler** [comment]
Same issue here, with same ubuntu version and same gst-inspect output.

**Comment by gkarsay** [comment]
I'm sorry for not being responsive for more than a month. This year is kind of special and brought some unexpected challenges. Anyway, try to stay in good spirits! – Now enough of this inappropriate non-technical off-topic meta-talk, lets go back to the current issue.

The problem is that a GStreamer plugin is not found that is shipped with Parlatype (so it should be there).

Possible reasons I can think of are:

### GStreamer can't find it

This should only happen if installed from source into a different prefix then `/usr`.
The plugin is called `libgstparlasphinx.so`, you can find it with this command and should get a similar result:
```
$ locate libgstparlasphinx.so
/usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgstparlasphinx.so
```
Try to launch parlatype specifying the plugin path, without the `/libgstparlasphinx.so` part like this:
```
GST_PLUGIN_PATH=/usr/lib/x86_64-linux-gnu/gstreamer-1.0 parlatype
```

### Plugin is blacklisted

There are several possible reasons. If it is blacklisted, I think `gst-inspect-1.0` would tell so, as already posted:
```
gst-inspect-1.0 /usr/lib/x86_64-linux-gnu/gstreamer-1.0/libgstparlasphinx.so
```
Or to see blacklisted plugins:
```
gst-inspect-1.0 -b
```
A reason I know of is a mismatch of the current GStreamer version and the GStreamer version that was used to compile the plugin. Do you have changed or upgraded your GStreamer installation? Did you upgrade your distro and did it work before and stopped working? I'm not sure, if this is only a problem, if the runtime is older or even if the runtime is newer.

## Other debugging

You can call Parlatype like this to have a bit more of information on the terminal, but not too detailed:
```
G_MESSAGES_DEBUG=all parlatype
```
You can get a lot of details with GStreamer debug levels, e.g. level 4:
```
GST_DEBUG=4 parlatype
```
Levels range from 1 to 9, 1 shows only errors, 2 warnings, 4 is info, 5 debug and that's probably more than enough. Hopefully you get some meaningful information.

That's all I can currently think of to solve the problem.

**Comment by donnekgit** [comment]
Hi.  Thanks for the notes.

The problem appears to be that the plugin is blacklisted.

```
GST_PLUGIN_PATH=/usr/lib/x86_64-linux-gnu/gstreamer-1.0 parlatype
```
gives the same error message.  Using gst-inspect gives no indication (the output is an analogue of that posted above) that the plugin is blacklisted, but `gst-inspect-1.0 -b` gives:
```
Blacklisted files:
  libgstparlasphinx.so
Total count: 1 blacklisted file
```

`G_MESSAGES_DEBUG=all parlatype` gives:
```
(process:199142): GLib-GIO-DEBUG: 11:40:41.552: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
(parlatype:199142): dbind-WARNING **: 11:40:41.578: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
(parlatype:199142): GLib-DEBUG: 11:40:41.579: unsetenv() is not thread-safe and should not be used after threads are created
(parlatype:199142): Gtk-DEBUG: 11:40:41.579: Connecting to session manager
(parlatype:199142): Gtk-DEBUG: 11:40:41.579: Failed to get the GNOME session proxy: The name org.gnome.SessionManager is not owned
(parlatype:199142): Gtk-DEBUG: 11:40:41.580: Failed to get the Xfce session proxy: The name org.xfce.SessionManager is not owned
```

However, this is a clean install of 20.04, with GStreamer coming from the standard repos.  I haven't recompiled or upgraded  GStreamer.  The only thing I can think of is that there is a problem upstream (ie that the plugin has been compiled against a different version of GStreamer than is in 20.04, or that some other multimedia app has unilaterally upgraded GStreamer.  I think the former is more likely than the latter, but I Am Not A Coder! :-)

**Comment by gkarsay** [comment]
At least we know it's blacklisted ... Did you get any interesting information from `GST_DEBUG=4 parlatype`?

Another thing to try is to regenerate GStreamer's registry: delete `~/.cache/gstreamer-1.0/registry.x86_64.bin`. Run `gst-inspect-1.0` to regenerate it. Is the plugin still blacklisted?

As I have installed Ubuntu 20.4 in a VM and it works for me, maybe remove the package and reinstall it? (Maybe something went wrong with the upgrade process. Did you upgrade from Bionic?)
`sudo apt purge parlatype libparlatype2; sudo apt install parlatype`

Btw package upstream is me, I packaged Parlatype for Debian and that is taken without modifications by Ubuntu.

**Comment by donnekgit** [comment]
Deleting the registry and regenerating it allows Parlatype to open.  Thanks.

I'd already tried deleting and reinstalling - that doesn't work.

Since others are having the same issue, there's maybe some issue with (a fresh bare-metal install of) 20.04 that doesn't show in a VM?

**Comment by gkarsay** [comment]
As a preliminary summary:
A **workaround for this problem** is
```
rm ~/.cache/gstreamer-1.0/registry.x86_64.bin
gst-inspect-1.0
```
On Debian/Ubuntu the command `gst-inspect-1.0` is provided by package `gstreamer1.0-tools`.

I'm still investigating the cause.

**Comment by gkarsay** [comment]
I haven't found the cause for this but fixed it in 3.0 by making Parlasphinx an internal, compiled in plugin, so that there should be no problems with GStreamer and the registry anymore.
